sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/TableProvider.java

```diff
@@ -36,26 +35,12 @@
 public interface TableProvider {
 
   /**
-   * Return a {@link Table} instance to do read/write with user-specified options.
+   * Return a {@link Table} instance with the given table options to do read/write.
+   * Implementations should infer the table schema and partitioning.
    *
    * @param options the user-specified options that can identify a table, e.g. file path, Kafka
    *                topic name, etc. It's an immutable case-insensitive string-to-string map.
    */
+  // TODO: this should take a Map<String, String> as table properties.
```

Comment:
If this method is left, then all of the changes to file sources can be avoided in this commit. That would make it considerably smaller.

Reply:
Another option for avoiding this is to separate the schema and partition inference from the `getTable` method. In that case, `TableProvider` would expose `inferSchema(CaseInsensitiveStringMap)` and `inferPartitioning(CaseInsensitiveStringMap)`. Then a single `getTable` call could be used.

```java
interface TableProvider {
  StructType inferSchema(CaseInsensitiveStringMap options);
  Transform[] inferPartitioning(CaseInsensitiveStringMap options);
  Table buildTable(StructType schema, Transform[] partitioning, Map<String, String> properties);
}
```

What do you think?

Reply:
SGTM. Shall we do it in this PR? We need to update all the `TableProvider` implementations, i.e. file source v2, streaming source v2, testing v2 sources.

Reply:
No, I think we should revert the changes to all the sources in this commit and continue to use the original method. In this PR, let's add the method to build a table and add support to the v2 session catalog. We can change to infer before calling that method in a follow-up.

Reply:
I did a quick try, and found one problem: it's very likely that a data source needs to infer schema and partition together, e.g. file source. If we have 2 separated method `inferSchema` and `inferPartitioning`, it may force the data source to do expensive inference twice.

Reply:
Can't we do `inferPartitioning` first and then pass that into `inferSchema`?

Reply:
Then API is limited to the assumption that schema inference depends on partition inference. This is true for file source, but I'm not sure if it's true to all data sources.

Shall we stick to the current proposal that has 2 `getTable` methods?

Reply:
It turns out that this is more complicated than I thought.

For file source, we need to list the files to scan. If we need to infer the schema/partitioning, we will list the files a little earlier, and reuse the listed files when scan.

That said, we need to either make `TableProvider` stateful, so that file source can keep the listed files after schema inference. Or we need to re-design the API so that it can carry states after schema inference.

Maybe we should move on with 2 `getTable` methods as it is in this PR.

Reply:
A TableProvider could use a static cache. We do this for our Hive client pool in Iceberg.

I don't think that partition inference needs to scan the entire file system tree. That seems needlessly expensive to me, when mixed depth partitions are not allowed. Inference should find the first live data file and use its path to determine the partitioning.

In any case, I think that the plan to add `inferSchema` and `inferPartitioning` is fine, it is just fixing the implementation that needs to be done. Schema and partitioning inference is known to be expensive, so I'm not sure how much sense it makes to try to over-optimize here.

Reply:
> I don't think that partition inference needs to scan the entire file system tree.

Spark needs to do it to get all the partition values and infer the schema. This is an existing feature that Spark can infer a common type for partition values with different types. The same applies to schema inference as well. Spark can read parquet files of different but compatible schema, so Spark must read all the files to infer the schema.

Can you share more about the static cache? Do you mean a global cache that maps a directory to its listed files?

Reply:
My point was that Spark needs to infer the partitioning of the table, not exhaustively list directories. This can be done more quickly than in the current implementation, by listing all files later and just getting the directory structure for `inferPartititoning`.

The static cache I'm talking about is a cache of metastore connections, not files. In this case, you could build your file list for a location and cache that for some period of time, using it for partition and schema inference, as well as for the `FileIndex` in the table you created. Caching would also help consistency because the same files would be in all versions of the table loaded within some period of time (and could be refreshed, of course). But, these concerns shouldn't affect the API.